pyro/distributions/score_parts.py

```diff
@@ -25,6 +25,6 @@ def scale_and_mask(self, scale=1.0, mask=None):
         :type mask: torch.BoolTensor or None
         """
         log_prob = scale_and_mask(self.log_prob, scale, mask)
-        score_function = self.score_function  # not scaled
+        score_function = scale_and_mask(self.score_function, 1.0, mask)  # not scaled
```

Comment:
Can you point out exactly the scenario that is fixed by this one line change?  IIRC, `score_function` would always be multiplied by another tensor that is masked, so the mask here would be redundant.

Reply:
Here it's again for consistency. When a variable is partially observed, some parameters are created for the missing observations, and some dummy parameters are created for the non-missing ones. When `has_rsample` is true, the dummy parameters are not updated: they retain their initial values because they do not contribute to the gradient. When `has_rsample` is false, the gradient "leaks" through this line and the dummy parameters are updated during learning. (but I found that inference on the non-dummy parameters was correct in the cases I checked). As above, this line does not really fix any bug, it just tries to make the behavior consistent between `has_rsample = True` and `has_rsample = False`.

Reply:
Thanks for explaining, I think I better understand now.